File: Backend/testRaptor.js
// testRaptor.js
const DataLoader = require("./fileLoader");
const { RAPTOR, minutesToTime, reconstructPath, timeToMinutes } = require("./RAPTOR");

// ADD THE DEBUG FUNCTION HERE
function debugJourneyPlanning(sourceStop, targetStop, departureTime, stops, trips, stopToRoutes, loader) {
    console.log(`\n🔍 COMPREHENSIVE DEBUG`);
    console.log(`Source: ${sourceStop} (${loader.getStopNameById(sourceStop)})`);
    console.log(`Target: ${targetStop} (${loader.getStopNameById(targetStop)})`);
    console.log(`Departure: ${departureTime}`);
    
    // 1. Check if stops exist
    console.log(`\n1️⃣ STOP VALIDATION`);
    console.log(`Source stop ${sourceStop} exists: ${sourceStop in stopToRoutes}`);
    console.log(`Target stop ${targetStop} exists: ${targetStop in stopToRoutes}`);
    
    // 2. Check routes from source
    console.log(`\n2️⃣ ROUTES FROM SOURCE`);
    const sourceRoutes = stopToRoutes[sourceStop] || [];
    console.log(`Routes from source: ${sourceRoutes.join(', ')}`);
    
    // 3. Find ALL trips that serve the source stop at or after departure time
    console.log(`\n3️⃣ ALL VALID TRIPS FROM SOURCE`);
    const departureMinutes = timeToMinutes(departureTime);
    
    sourceRoutes.forEach(routeID => {
        console.log(`\n--- Route ${routeID} ---`);
        const routeTrips = Object.values(trips).filter(t => t.route === routeID);
        
        const validTrips = [];
        routeTrips.forEach(trip => {
            const sourceStopInTrip = trip.times.find(t => t.stopID === sourceStop);
            if (sourceStopInTrip) {
                const tripDepartureTime = timeToMinutes(sourceStopInTrip.time);
                const isValid = tripDepartureTime >= departureMinutes;
                
                console.log(`Trip ${trip.tripID}: departs ${sourceStopInTrip.time} (${isValid ? '✅ VALID' : '❌ TOO EARLY'})`);
                
                if (isValid) {
                    validTrips.push({
                        tripID: trip.tripID,
                        departureTime: sourceStopInTrip.time,
                        stops: trip.times.map(t => `${loader.getStopNameById(t.stopID)}@${t.time}`).join(' → ')
                    });
                }
            }
        });
        
        console.log(`Valid trips: ${validTrips.length}`);
        validTrips.forEach(trip => {
            console.log(`  🚆 ${trip.tripID} @ ${trip.departureTime}`);
            console.log(`     Route: ${trip.stops}`);
            
            // Check if this trip reaches target directly
            const tripData = trips[trip.tripID];
            const targetInTrip = tripData.times.find(t => t.stopID === targetStop);
            if (targetInTrip) {
                console.log(`     🎯 DIRECT CONNECTION TO TARGET @ ${targetInTrip.time}`);
            }
        });
    });
    
    // 4. Check what routes serve the target
    console.log(`\n4️⃣ ROUTES TO TARGET`);
    const targetRoutes = stopToRoutes[targetStop] || [];
    console.log(`Routes serving target: ${targetRoutes.join(', ')}`);
    
    // 5. Find direct connections
    console.log(`\n5️⃣ DIRECT CONNECTIONS`);
    const commonRoutes = sourceRoutes.filter(r => targetRoutes.includes(r));
    console.log(`Common routes (direct connection possible): ${commonRoutes.join(', ')}`);
    
    commonRoutes.forEach(routeID => {
        console.log(`\nDirect trips on route ${routeID}:`);
        const routeTrips = Object.values(trips).filter(t => t.route === routeID);
        
        routeTrips.forEach(trip => {
            const sourceStopInTrip = trip.times.find(t => t.stopID === sourceStop);
            const targetStopInTrip = trip.times.find(t => t.stopID === targetStop);
            
            if (sourceStopInTrip && targetStopInTrip) {
                const sourceDep = timeToMinutes(sourceStopInTrip.time);
                const targetArr = timeToMinutes(targetStopInTrip.time);
                const isValidTime = sourceDep >= departureMinutes;
                const isForward = targetArr > sourceDep;
                
                console.log(`Trip ${trip.tripID}: ${sourceStopInTrip.time} → ${targetStopInTrip.time} ${isValidTime && isForward ? '✅' : '❌'}`);
                if (isValidTime && isForward) {
                    console.log(`  🎯 PERFECT DIRECT TRIP!`);
                }
            }
        });
    });
    
    // 6. Sample a few trips to check data integrity
    console.log(`\n6️⃣ DATA INTEGRITY CHECK`);
    const sampleTrips = Object.values(trips).slice(0, 3);
    sampleTrips.forEach(trip => {
        console.log(`Trip ${trip.tripID} sample:`, 
            trip.times.slice(0, 5).map(t => `${t.stopID}@${t.time}`).join(' → ')
        );
    });
}

// arguments: node testRaptor.js "CAPE TOWN" "WOODSTOCK" 04:30
const [,, sourceStopName, targetStopName, departureTime] = process.argv;

if (!sourceStopName || !targetStopName || !departureTime) {
    console.log("Usage: node testRaptor.js <sourceStop> <targetStop> <HH:MM>");
    process.exit(1);
}

const loader = new DataLoader();
const filePath = "C:\\Users\\Phalanndwa Zwivhuya\\Documents\\School stuff\\computer science\\3rd year\\2025\\CSC3003S\\Data\\CapeTownTransitData\\train-schedules-2014\\S1.csv";
const data = loader.loadFromFile(filePath);


// BUILD stopToRoutes mapping (you might be missing this!)
const stopToRoutes = {};
Object.values(data.trips).forEach(trip => {
    trip.times.forEach(stopTime => {
        const stopID = stopTime.stopID;
        const routeID = trip.route;
        
        if (!stopToRoutes[stopID]) {
            stopToRoutes[stopID] = [];
        }
        if (!stopToRoutes[stopID].includes(routeID)) {
            stopToRoutes[stopID].push(routeID);
        }
    });
});

const sourceStop = loader.findStopByName(sourceStopName.toUpperCase());
const targetStop = loader.findStopByName(targetStopName.toUpperCase());

// Add this right after loading the data in testRaptor.js
console.log("\n🔍 DEBUGGING TRIP 3535");
const trip3535 = data.trips['3535'];
if (trip3535) {
    console.log("Trip 3535 data:", trip3535);
    console.log("All stop times:");
    trip3535.times.forEach((stopTime, index) => {
        console.log(`  ${index}: Stop ${stopTime.stopID} (${loader.getStopNameById(stopTime.stopID)}) at ${stopTime.time}`);
    });
    
    // Find Cape Town specifically
    const capeStopTime = trip3535.times.find(t => t.stopID === sourceStop);
    if (capeStopTime) {
        console.log(`Cape Town departure: ${capeStopTime.time}`);
    } else {
        console.log("Cape Town not found in this trip!");
    }
} else {
    console.log("Trip 3535 not found in data!");
}

if (sourceStop === undefined || targetStop === undefined) {
    console.error("❌ Could not find source or target stop");
    console.log("Available stops:", loader.getAvailableStops().slice(0, 10).join(", "), "...");
    process.exit(1);
}

console.log(`\n=== Planning Journey from ${sourceStopName} to ${targetStopName} at ${departureTime} ===`);

// ADD DEBUG CALL HERE
debugJourneyPlanning(sourceStop, targetStop, departureTime, data.stops, data.trips, stopToRoutes, loader);

// THEN RUN RAPTOR
const result = RAPTOR(sourceStop, targetStop, departureTime, data.stops, data.trips, stopToRoutes);

if (result.earliestArrival[targetStop] === Infinity) {
    console.log("❌ No path found");
} else {
    console.log("✅ Earliest arrival:", minutesToTime(result.earliestArrival[targetStop]));
    const path = reconstructPath(result.predecessor, sourceStop, targetStop, loader, data.trips);
    path.forEach((step, i) => {
        console.log(`${i + 1}. ${step.stopName} at ${step.time} (Trip: ${step.tripID})`);
    });
}

File: Backend/src/TestRaptor.java
package backend;

import java.util.*;

public class TestRaptor {
// Debug function
// Remove-Item bin\* -Recurse -Force to clean the bin after compiling 
// javac -d bin src/*.java - to compile again 
// java -cp bin src.TestRaptor MUTUAL "CAPE TOWN" 07:21 to test the times on each occaaion 






    public static void debugJourneyPlanning(
            int sourceStop,
            int targetStop,
            String departureTime,
            Map<String, Integer> stops,
            Map<String, Trip> trips,
            Map<Integer, List<String>> stopToRoutes,
            DataLoader loader
    ) {
        System.out.println("\n🔍 COMPREHENSIVE DEBUG");
        System.out.println("Source: " + sourceStop + " (" + loader.getStopNameById(sourceStop) + ")");
        System.out.println("Target: " + targetStop + " (" + loader.getStopNameById(targetStop) + ")");
        System.out.println("Departure: " + departureTime);

        // 1. Check if stops exist
        System.out.println("\n1️⃣ STOP VALIDATION");
        System.out.println("Source stop exists: " + stopToRoutes.containsKey(sourceStop));
        System.out.println("Target stop exists: " + stopToRoutes.containsKey(targetStop));

        // 2. Routes from source
        System.out.println("\n2️⃣ ROUTES FROM SOURCE");
        List<String> sourceRoutes = stopToRoutes.getOrDefault(sourceStop, new ArrayList<>());
        System.out.println("Routes from source: " + String.join(", ", sourceRoutes));

        // 3. All valid trips from source
        System.out.println("\n3️⃣ ALL VALID TRIPS FROM SOURCE");
        int departureMinutes = Raptor.timeToMinutes(departureTime);

        for (String routeID : sourceRoutes) {
            System.out.println("\n--- Route " + routeID + " ---");
            List<Trip> routeTrips = new ArrayList<>();
            for (Trip t : trips.values()) {
                if (t.route.equals(routeID)) {
                    routeTrips.add(t);
                }
            }

            List<Trip> validTrips = new ArrayList<>();
            for (Trip trip : routeTrips) {
                StopTime sourceStopInTrip = trip.times.stream()
                        .filter(t -> t.stopID == sourceStop)
                        .findFirst().orElse(null);

                if (sourceStopInTrip != null) {
                    int tripDepartureTime = Raptor.timeToMinutes(sourceStopInTrip.time);
                    boolean isValid = tripDepartureTime >= departureMinutes;

                    System.out.println("Trip " + trip.tripID + ": departs " +
                            sourceStopInTrip.time + " (" + (isValid ? "✅ VALID" : "❌ TOO EARLY") + ")");

                    if (isValid) {
                        validTrips.add(trip);
                    }
                }
            }

            System.out.println("Valid trips: " + validTrips.size());
            for (Trip trip : validTrips) {
                StringBuilder sb = new StringBuilder();
                for (StopTime st : trip.times) {
                    sb.append(loader.getStopNameById(st.stopID)).append("@").append(st.time).append(" → ");
                }
                String routeStops = sb.substring(0, sb.length() - 3);
                System.out.println("  🚆 " + trip.tripID + " @ " + trip.times.get(0).time);
                System.out.println("     Route: " + routeStops);

                // Direct connection check
                StopTime targetInTrip = trip.times.stream()
                        .filter(t -> t.stopID == targetStop)
                        .findFirst().orElse(null);
                if (targetInTrip != null) {
                    System.out.println("     🎯 DIRECT CONNECTION TO TARGET @ " + targetInTrip.time);
                }
            }
        }

        // 4. Routes to target
        System.out.println("\n4️⃣ ROUTES TO TARGET");
        List<String> targetRoutes = stopToRoutes.getOrDefault(targetStop, new ArrayList<>());
        System.out.println("Routes serving target: " + String.join(", ", targetRoutes));

        // 5. Direct connections
        System.out.println("\n5️⃣ DIRECT CONNECTIONS");
        List<String> commonRoutes = new ArrayList<>(sourceRoutes);
        commonRoutes.retainAll(targetRoutes);
        System.out.println("Common routes: " + String.join(", ", commonRoutes));

        for (String routeID : commonRoutes) {
            System.out.println("\nDirect trips on route " + routeID + ":");
            for (Trip trip : trips.values()) {
                if (!trip.route.equals(routeID)) continue;

                StopTime sourceStopInTrip = trip.times.stream()
                        .filter(t -> t.stopID == sourceStop)
                        .findFirst().orElse(null);
                StopTime targetStopInTrip = trip.times.stream()
                        .filter(t -> t.stopID == targetStop)
                        .findFirst().orElse(null);

                if (sourceStopInTrip != null && targetStopInTrip != null) {
                    int sourceDep = Raptor.timeToMinutes(sourceStopInTrip.time);
                    int targetArr = Raptor.timeToMinutes(targetStopInTrip.time);
                    boolean isValidTime = sourceDep >= departureMinutes;
                    boolean isForward = targetArr > sourceDep;

                    System.out.println("Trip " + trip.tripID + ": " +
                            sourceStopInTrip.time + " → " + targetStopInTrip.time +
                            (isValidTime && isForward ? " ✅" : " ❌"));
                }
            }
        }

        // 6. Data integrity sample
        System.out.println("\n6️⃣ DATA INTEGRITY CHECK");
        int count = 0;
        for (Trip trip : trips.values()) {
            System.out.print("Trip " + trip.tripID + " sample: ");
            for (int i = 0; i < Math.min(5, trip.times.size()); i++) {
                StopTime st = trip.times.get(i);
                System.out.print(st.stopID + "@" + st.time + " → ");
            }
            System.out.println();
            if (++count >= 3) break;
        }
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 3) {
            System.out.println("Usage: java TestRaptor <sourceStop> <targetStop> <HH:MM>");
            return;
        }

        String sourceStopName = args[0];
        String targetStopName = args[1];
        String departureTime = args[2];

        DataLoader loader = new DataLoader();
        String filePath = "C:\\Users\\Phalanndwa Zwivhuya\\Documents\\School stuff\\computer science\\3rd year\\2025\\CSC3003S\\Data\\CapeTownTransitData\\train-schedules-2014\\S1.csv";
        List<String[]> rows = loader.loadCSV(filePath);
        loader.buildData(rows);

        // Build stopToRoutes (already built in DataLoader, but let's copy it)
        Map<Integer, List<String>> stopToRoutes = loader.stopToRoutes;

        Integer sourceStop = loader.findStopByName(sourceStopName.toUpperCase());
        Integer targetStop = loader.findStopByName(targetStopName.toUpperCase());

        if (sourceStop == null || targetStop == null) {
            System.out.println("❌ Could not find source or target stop");
            System.out.println("Available stops (sample): " + loader.getAvailableStops());
            return;
        }

        System.out.println("\n=== Planning Journey from " +
                sourceStopName + " to " + targetStopName + " at " + departureTime + " ===");

        // Debugging
        debugJourneyPlanning(sourceStop, targetStop, departureTime,
                loader.stops, loader.trips, stopToRoutes, loader);

        // Run RAPTOR
        Result result = Raptor.runRaptor(
                sourceStop, targetStop, departureTime,
                loader.stops, loader.trips, stopToRoutes
        );

        if (result.earliestArrival[targetStop] == Integer.MAX_VALUE) {
            System.out.println("❌ No path found");
        } else {
            System.out.println("✅ Earliest arrival: " +
                    Raptor.minutesToTime(result.earliestArrival[targetStop]));
            List<PathStep> path = Raptor.reconstructPath(
                    result.predecessor, sourceStop, targetStop, loader, loader.trips
            );
            for (int i = 0; i < path.size(); i++) {
                System.out.println((i + 1) + ". " + path.get(i));
            }
        }
    }
}
